File: backend/app/domain/ledger.py
"""Счетоводна книга на злато-обезпечената сметка.

Огледало на ``src/core/ledger.ts``. Разликата в ролите е съществена:
клиентът смята, за да ПОКАЖЕ какво ще стане, а този модул решава какво
наистина става. Затова тук няма нито едно допускане, което да идва от
клиента — сумата се валидира наново, курсът се взима от сървърния поток,
а наличността се чете от базата.

Функциите са чисти: вход → резултат, без достъп до база или време. Записът
в базата се прави от слоя с услугите, който ги вика.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum

from .gold import GoldQuote
from .money import MICROGRAMS_PER_GRAM, fee_of


class TxKind(str, Enum):
    """Вид движение по сметката."""

    TOPUP = "topup"
    """Зареждане с евро → покупка на злато."""

    CARD = "card"
    """Плащане с картата → продажба на злато."""

    SELL = "sell"
    """Ръчна продажба на злато → изплащане в евро."""


class Fees:
    """Такси на продукта.

    Зареждането е без такса — платформата печели от спреда. Плащането с
    картата се таксува само над безплатния месечен оборот.
    """

    TOPUP = 0.0
    CARD_OVER_LIMIT = 0.005
    FREE_MONTHLY_CARD_VOLUME = 100_000  # 1 000,00 EUR
    SELL = 0.0025


#: Лимити за зареждане в евроцентове.
TOPUP_MIN = 10_00
TOPUP_MAX = 1_000_000  # 10 000,00 EUR


class LedgerError(Exception):
    """Отказ по бизнес причина, а не по техническа.

    Носи машинно четим ``code``, за да може API слоят да го преведе в
    HTTP отговор, без да разчита на текста на съобщението.
    """

    def __init__(self, code: str, message: str, **details: object) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.details = details


@dataclass(frozen=True, slots=True)
class Movement:
    """Резултат от една сделка: какво се записва и как се мести салдото.

    Не пипа база — слоят с услугите го превръща в ред в таблицата.
    """

    kind: TxKind
    title: str
    amount: int
    """Брутна сума в евроцентове."""

    fee: int
    """Удържана такса в евроцентове."""

    gold_delta: int
    """Злато, купено (+) или продадено (−), в микрограмове."""

    price_per_gram: int
    """Курсът, приложен към сделката."""


def gold_for_amount(amount: int, price_per_gram: int) -> int:
    """Колко микрограма злато се получават за дадена сума.

    Умножава се преди делението, за да е делението единствената операция със
    закръгляне — същата формула като в клиента, за да съвпадат числата.
    Закръгля се НАДОЛУ: платформата не издава злато, което не е купила.
    """
    if price_per_gram <= 0:
        return 0
    return (amount * MICROGRAMS_PER_GRAM) // price_per_gram


def amount_for_gold(gold: int, price_per_gram: int) -> int:
    """Колко евроцента струва дадено количество злато."""
    return round((gold * price_per_gram) / MICROGRAMS_PER_GRAM)


def gold_needed_for_amount(amount: int, price_per_gram: int) -> int:
    """Колко злато трябва да се продаде, за да се покрие сума.

    Закръгля се НАГОРЕ: покритието трябва да е пълно, иначе сметката остава
    с дефицит от част от цента.
    """
    if price_per_gram <= 0:
        raise LedgerError("invalid_quote", "Невалидна котировка.")
    return math.ceil((amount * MICROGRAMS_PER_GRAM) / price_per_gram)


def month_start(at: float) -> float:
    """Начало на календарния месец, в който попада моментът (UTC)."""
    moment = datetime.fromtimestamp(at, tz=timezone.utc)
    return moment.replace(
        day=1, hour=0, minute=0, second=0, microsecond=0
    ).timestamp()


def card_fee(amount: int, card_volume_this_month: int) -> int:
    """Такса за плащане с карта: само върху частта над безплатния лимит."""
    remaining_free = max(Fees.FREE_MONTHLY_CARD_VOLUME - card_volume_this_month, 0)
    chargeable = max(amount - remaining_free, 0)
    return fee_of(chargeable, Fees.CARD_OVER_LIMIT)


def plan_topup(amount: int, quote: GoldQuote) -> Movement:
    """Зареждане: купува се злато по ask цената."""
    if amount < TOPUP_MIN:
        raise LedgerError(
            "amount_too_small",
            f"Минималната сума за зареждане е {TOPUP_MIN / 100:.2f} EUR.",
            min=TOPUP_MIN,
        )
    if amount > TOPUP_MAX:
        raise LedgerError(
            "amount_too_large",
            f"Максималната сума за зареждане е {TOPUP_MAX / 100:.2f} EUR.",
            max=TOPUP_MAX,
        )

    fee = fee_of(amount, Fees.TOPUP)
    bought = gold_for_amount(amount - fee, quote.ask_per_gram)

    return Movement(
        kind=TxKind.TOPUP,
        title="Зареждане и покупка на злато",
        amount=amount,
        fee=fee,
        gold_delta=bought,
        price_per_gram=quote.ask_per_gram,
    )


def plan_card_payment(
    amount: int,
    quote: GoldQuote,
    merchant: str,
    *,
    available_gold: int,
    card_volume_this_month: int,
) -> Movement:
    """Плащане с картата: продава се точно толкова злато, колкото покрива
    сумата плюс таксата, по bid цената в момента на авторизация."""
    if amount <= 0:
        raise LedgerError("amount_too_small", "Сумата трябва да е положителна.", min=1)

    fee = card_fee(amount, card_volume_this_month)
    required_gold = gold_needed_for_amount(amount + fee, quote.bid_per_gram)

    if required_gold > available_gold:
        raise LedgerError(
            "insufficient_gold",
            "Наличното злато не покрива плащането.",
            required_gold=required_gold,
            available_gold=available_gold,
        )

    return Movement(
        kind=TxKind.CARD,
        title=merchant,
        amount=amount,
        fee=fee,
        gold_delta=-required_gold,
        price_per_gram=quote.bid_per_gram,
    )


def plan_sell(gold: int, quote: GoldQuote, *, available_gold: int) -> Movement:
    """Ръчна продажба на злато обратно в евро."""
    if gold <= 0:
        raise LedgerError("amount_too_small", "Количеството трябва да е положително.", min=1)
    if gold > available_gold:
        raise LedgerError(
            "insufficient_gold",
            "Наличното злато не стига за тази продажба.",
            required_gold=gold,
            available_gold=available_gold,
        )

    gross = amount_for_gold(gold, quote.bid_per_gram)
    fee = fee_of(gross, Fees.SELL)

    return Movement(
        kind=TxKind.SELL,
        title="Продажба на злато",
        amount=gross,
        fee=fee,
        gold_delta=-gold,
        price_per_gram=quote.bid_per_gram,
    )

File: src/core/ledger.ts
import {
  feeOf,
  type Cents,
  type Micrograms,
  MICROGRAMS_PER_GRAM,
} from './money';
import type { GoldQuote } from './goldPrice';

/**
 * Счетоводна книга на злато-обезпечената сметка.
 *
 * Модел на продукта: клиентът НЕ държи евро салдо. Всяко зареждане купува
 * злато веднага; всяко плащане с картата продава точно толкова злато,
 * колкото трябва, за да покрие сумата в момента на авторизация.
 *
 * Всички функции тук са чисти (pure): взимат състояние + събитие и връщат
 * ново състояние. Затова са тестваеми без UI и се преизползват дословно
 * от React Native клиента.
 */

export type TxKind =
  /** Зареждане с евро → покупка на злато. */
  | 'topup'
  /** Плащане с картата → продажба на злато. */
  | 'card'
  /** Ръчна продажба на злато → изплащане в евро. */
  | 'sell';

export interface Transaction {
  readonly id: string;
  readonly kind: TxKind;
  readonly at: number;
  /** Описание за изписване в историята. */
  readonly title: string;
  /** Брутна сума в евроцентове (това, което клиентът внася или плаща). */
  readonly amount: Cents;
  /** Удържана такса в евроцентове. */
  readonly fee: Cents;
  /** Злато, купено (+) или продадено (−) в микрограмове. */
  readonly goldDelta: Micrograms;
  /** Курсът за грам, приложен към сделката (евроцентове за грам). */
  readonly pricePerGram: Cents;
}

export interface Account {
  /** Наличност злато в микрограмове. */
  readonly gold: Micrograms;
  /** Сума на всички зареждания в евроцентове — база за пресмятане на P&L. */
  readonly deposited: Cents;
  /** Сума на всички изтеглени/похарчени евро. */
  readonly withdrawn: Cents;
  /** Сума на платените такси — показва се прозрачно в демото. */
  readonly feesPaid: Cents;
  readonly transactions: readonly Transaction[];
}

export const EMPTY_ACCOUNT: Account = {
  gold: 0,
  deposited: 0,
  withdrawn: 0,
  feesPaid: 0,
  transactions: [],
};

/**
 * Такси на продукта.
 *
 * Зареждането е без такса — платформата печели от спреда (виж `SPREAD`).
 * Плащането с картата също минава по спред; допълнителната такса се начислява
 * само над безплатния месечен лимит, както при повечето мулти-валутни карти.
 */
export const FEES = {
  /** Такса при зареждане (върху брутната сума). */
  topup: 0,
  /** Такса при плащане с карта над безплатния лимит. */
  cardOverLimit: 0.005,
  /** Безплатен месечен оборот по картата в евроцентове (1 000,00 EUR). */
  freeMonthlyCardVolume: 100_000 as Cents,
  /** Такса при ръчна продажба на злато обратно в евро. */
  sell: 0.0025,
} as const;

/** Минимално и максимално зареждане в евроцентове. */
export const TOPUP_MIN: Cents = 10_00;
export const TOPUP_MAX: Cents = 1_000_000; // 10 000,00 EUR

let sequence = 0;
const nextId = (): string => {
  sequence += 1;
  return `tx_${sequence.toString(36)}_${Date.now().toString(36)}`;
};

/**
 * Колко микрограма злато се получават за дадена сума по дадена цена за грам.
 *
 * Умножаваме ПРЕДИ да разделим: така делението е единствената операция със
 * закръгляне и резултатът е същият като на сървъра, който смята с цели числа.
 * Междинната стойност е под 2^53, защото зареждането е с горен лимит.
 */
export function goldForAmount(amount: Cents, pricePerGram: Cents): Micrograms {
  if (pricePerGram <= 0) return 0;
  // Закръгляме НАДОЛУ — платформата никога не издава злато, което не е купила.
  return Math.floor((amount * MICROGRAMS_PER_GRAM) / pricePerGram);
}

/** Колко евроцента струва дадено количество злато по дадена цена за грам. */
export function amountForGold(gold: Micrograms, pricePerGram: Cents): Cents {
  return Math.round((gold * pricePerGram) / MICROGRAMS_PER_GRAM);
}

/** Пазарна стойност на сметката, ако се ликвидира сега (по bid цената). */
export function portfolioValue(account: Account, quote: GoldQuote): Cents {
  return amountForGold(account.gold, quote.bidPerGram);
}

/**
 * Нереализиран резултат: колко струва златото сега спрямо нетно вложеното.
 * Отрицателна стойност е напълно възможна — златото пада, както и се качва.
 */
export function unrealisedPnl(account: Account, quote: GoldQuote): Cents {
  const invested = account.deposited - account.withdrawn;
  return portfolioValue(account, quote) - invested;
}

/** Оборот по картата за текущия календарен месец — база за безплатния лимит. */
export function cardVolumeThisMonth(account: Account, now: number): Cents {
  const start = new Date(now);
  start.setDate(1);
  start.setHours(0, 0, 0, 0);
  const from = start.getTime();
  return account.transactions
    .filter((tx) => tx.kind === 'card' && tx.at >= from)
    .reduce((sum, tx) => sum + tx.amount, 0);
}

export type LedgerError =
  | { code: 'amount_too_small'; min: Cents }
  | { code: 'amount_too_large'; max: Cents }
  | { code: 'insufficient_gold'; requiredGold: Micrograms; availableGold: Micrograms };

export type LedgerResult =
  | { ok: true; account: Account; transaction: Transaction }
  | { ok: false; error: LedgerError };

/**
 * Зареждане: клиентът внася евро, платформата купува злато по ask цената.
 */
export function topUp(
  account: Account,
  amount: Cents,
  quote: GoldQuote,
  now: number = Date.now(),
): LedgerResult {
  if (amount < TOPUP_MIN) {
    return { ok: false, error: { code: 'amount_too_small', min: TOPUP_MIN } };
  }
  if (amount > TOPUP_MAX) {
    return { ok: false, error: { code: 'amount_too_large', max: TOPUP_MAX } };
  }

  const fee = feeOf(amount, FEES.topup);
  const net = amount - fee;
  const bought = goldForAmount(net, quote.askPerGram);

  const transaction: Transaction = {
    id: nextId(),
    kind: 'topup',
    at: now,
    title: 'Зареждане и покупка на злато',
    amount,
    fee,
    goldDelta: bought,
    pricePerGram: quote.askPerGram,
  };

  return {
    ok: true,
    transaction,
    account: {
      ...account,
      gold: account.gold + bought,
      deposited: account.deposited + amount,
      feesPaid: account.feesPaid + fee,
      transactions: [transaction, ...account.transactions],
    },
  };
}

/**
 * Плащане с картата: продава се точно толкова злато, колкото покрива
 * сумата плюс таксата, по bid цената в момента на авторизация.
 */
export function cardPayment(
  account: Account,
  amount: Cents,
  quote: GoldQuote,
  merchant: string,
  now: number = Date.now(),
): LedgerResult {
  if (amount <= 0) {
    return { ok: false, error: { code: 'amount_too_small', min: 1 } };
  }

  // Таксата се начислява само върху частта над безплатния месечен лимит.
  const usedThisMonth = cardVolumeThisMonth(account, now);
  const remainingFree = Math.max(FEES.freeMonthlyCardVolume - usedThisMonth, 0);
  const chargeable = Math.max(amount - remainingFree, 0);
  const fee = feeOf(chargeable, FEES.cardOverLimit);

  const total = amount + fee;
  const requiredGold = Math.ceil((total * MICROGRAMS_PER_GRAM) / quote.bidPerGram);

  if (requiredGold > account.gold) {
    return {
      ok: false,
      error: {
        code: 'insufficient_gold',
        requiredGold,
        availableGold: account.gold,
      },
    };
  }

  const transaction: Transaction = {
    id: nextId(),
    kind: 'card',
    at: now,
    title: merchant,
    amount,
    fee,
    goldDelta: -requiredGold,
    pricePerGram: quote.bidPerGram,
  };

  return {
    ok: true,
    transaction,
    account: {
      ...account,
      gold: account.gold - requiredGold,
      withdrawn: account.withdrawn + total,
      feesPaid: account.feesPaid + fee,
      transactions: [transaction, ...account.transactions],
    },
  };
}

/**
 * Ръчна продажба на злато обратно в евро (например към банкова сметка).
 */
export function sellGold(
  account: Account,
  gold: Micrograms,
  quote: GoldQuote,
  now: number = Date.now(),
): LedgerResult {
  if (gold <= 0) {
    return { ok: false, error: { code: 'amount_too_small', min: 1 } };
  }
  if (gold > account.gold) {
    return {
      ok: false,
      error: {
        code: 'insufficient_gold',
        requiredGold: gold,
        availableGold: account.gold,
      },
    };
  }

  const gross = amountForGold(gold, quote.bidPerGram);
  const fee = feeOf(gross, FEES.sell);

  const transaction: Transaction = {
    id: nextId(),
    kind: 'sell',
    at: now,
    title: 'Продажба на злато',
    amount: gross,
    fee,
    goldDelta: -gold,
    pricePerGram: quote.bidPerGram,
  };

  return {
    ok: true,
    transaction,
    account: {
      ...account,
      gold: account.gold - gold,
      withdrawn: account.withdrawn + gross - fee,
      feesPaid: account.feesPaid + fee,
      transactions: [transaction, ...account.transactions],
    },
  };
}
